modules/codegen/src/main/scala/com/twilio/guardrail/generators/circe/protocol/CirceProtocolGenerator.scala

```diff
@@ -313,45 +324,59 @@ object CirceProtocolGenerator {
-        val code = q"""object ${Term.Name(clsName)} extends AutoDerivation {
-            ..${List(discriminator, encoder, decoder)}
+        val code = q"""object ${Term.Name(clsName)} {
+             val discriminator:String = ${Lit.String(discriminator)}
+            ..${List(encoder, decoder)}
           }
           """
         Target.pure(code)
 
-      case RenderDiscriminator(discriminator) =>
-        val code = q"""implicit val configuration: Configuration = Configuration.default.withDiscriminator($discriminator)"""
-        Target.pure(code)
-
-      case DecodeADT(clsName, needCamelSnakeConversion) =>
-        val code = if (needCamelSnakeConversion) {
-          q"""implicit val decoder: Decoder[${Type.Name(clsName)}] =
-                  deriveDecoder[${Type.Name(clsName)}](io.circe.derivation.renaming.snakeCase)"""
-        } else {
-          q"""implicit val decoder: Decoder[${Type.Name(clsName)}] = deriveDecoder[${Type.Name(clsName)}]"""
-        }
+      case DecodeADT(clsName, children) =>
+        val childrenCases = children.map(
+          child => p"case ${Lit.String(child)} => c.as[${Type.Name(child)}]"
+        )
+        val code =
+          q"""implicit val decoder: Decoder[${Type.Name(clsName)}] = Decoder.instance(c =>
+                 c.downField(discriminator).as[String].flatMap {
+                   ..case $childrenCases
+                 }
+            )"""
         Target.pure(code)
 
-      case EncodeADT(clsName, needCamelSnakeConversion) =>
-        val code = if (needCamelSnakeConversion) {
-          q"""implicit val encoder: Encoder[${Type.Name(clsName)}] = deriveEncoder[${Type
-            .Name(clsName)}](io.circe.derivation.renaming.snakeCase)"""
-        } else {
-          q"""implicit val encoder: Encoder[${Type.Name(clsName)}] = deriveEncoder[${Type.Name(clsName)}]"""
-        }
-
+      case EncodeADT(clsName, children) =>
+        val childrenCases = children.map(
+          child => p"case e:${Type.Name(child)} => e.asJsonObject.add(discriminator, ${Lit.String(child)}.asJson).asJson"
+        )
+        val code =
+          q"""implicit val encoder: Encoder[${Type.Name(clsName)}] = Encoder.instance {
+              ..case $childrenCases
+          }"""
         Target.pure(code)
 
-      case RenderSealedTrait(className, terms, discriminator: String) =>
+      case RenderSealedTrait(className, terms, discriminator, parents) =>
+        val parentNameOpt = parents.headOption.map(_.clsName)
         val testTerms = terms
           .filter(_.name.value != discriminator)
           .map { t =>
-            q"""def ${Term.Name(t.name.value)} : ${t.decltpe.getOrElse(Type.Name("Any"))}"""
+            val tpe: Type = t.decltpe
+              .flatMap({
+                case tpe: Type => Some(tpe)
+                case x =>
+                  println(s"Unsure how to map ${x.structure}, please report this bug!")
+                  None
+              })
+              .get
```

Comment:
Instead of using `get`, this should be a `Target.error("<error message>")` instead of using `Target.pure` below

Reply:
It's copy-past from DecodeModel.
How should I do it? I mean. We have list of properties and if any one property is incorrect. Should I return `Target.error("<error message>")` or might I just skip the property?

Reply:
Definitely use `Target.error`. Dropping a property is not adhering to the protocol specification. It's important that we raise this error to the user quickly instead of having them hunt through the build logs to find out why.

I have a change to the sbt and Maven plugins that fails compilation if there are any errors, as currently we just fail to emit any code and continue compiling.

Reply:
None.get it's impossible situation. If we take a look at the part of code
```
(finalDeclType, finalDefaultValue) = Option(property.getRequired)
              .filterNot(_ == false)
              .fold[(Type, Option[Term])](
                (t"Option[${tpe}]", Some(defaultValue.fold[Term](q"None")(t => q"Option($t)")))
              )(Function.const((tpe, defaultValue)) _)
            term = param"${Term.Name(argName)}: ${finalDeclType}".copy(default = finalDefaultValue)
            dep  = rawDep.filterNot(_.value == clsName) // Filter out our own class name
```
we can to see `param"${Term.Name(argName)}: ${finalDeclType}".`. `Type` is necessary.